src/cares_wrap.cc

```diff
@@ -743,8 +753,14 @@ class QueryWrap : public AsyncWrap {
   }
 
   ChannelWrap* channel_;
+  ares_function* function_;
 };
 
+class QueryWrap : public BaseWrap {
+ public:
+  QueryWrap(ChannelWrap* channel, Local<Object> req_wrap_obj, ares_function* fn)
+      : BaseWrap(channel, req_wrap_obj, fn) {}
+};
```

Comment:
I think we don’t need the `BaseWrap` layer of indirection anymore, right?

Reply:
obvious, once you point it out 😁

removed. now how do we provide search domains, so that we can write some meaningful tests for `search: true`?

Reply:
@SirR4T I don’t know the feature well enough (yet?) to understand how to write good tests for it, but yes, I think we’re at that point generally. The [mock DNS module](https://github.com/nodejs/node/blob/master/test/common/README.md#dns-module) might help a bit?

Reply:
Seems to me that we will need to get the `ares_channel channel->ndomains` setup properly, for `ares_search()` to be meaningful (not default to using `ares_query()` instead).

As far as i can gather, there are [four ways of doing this](https://github.com/nodejs/node/blob/master/deps/cares/src/ares_init.c#L191-L211). Not sure which would be easiest to mock / replicate.

Reply:
strike that, i was going down the wrong rabbit hole.

Here's what's working, now: 

**Search for "en" fails, before any changes to `/etc/resolv.conf`:**
```
> dns.resolve4('en', {ttl: true, search: true}, (err, result) => { if (err) console.error(err); console.log(result); })
QueryReqWrap {
  bindingName: 'queryA',
  callback: [Function],
  hostname: 'en',
  oncomplete: [Function: onresolve],
  ttl: true,
  search: true,
  domain:
   Domain {
     domain: null,
     _events:
      { removeListener: [Function: updateExceptionCapture],
        newListener: [Function: updateExceptionCapture],
        error: [Function: debugDomainError] },
     _eventsCount: 3,
     _maxListeners: undefined,
     members: [] },
  channel:
   ChannelWrap {
     domain:
      Domain {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        members: [] } } }
> { Error: queryA ESERVFAIL en
    at QueryReqWrap.onresolve [as oncomplete] (dns.js:197:19)
  errno: 'ESERVFAIL',
  code: 'ESERVFAIL',
  syscall: 'queryA',
  hostname: 'en' }
```
Same with `nslookup`:
```
$ nslookup "en"
;; Got SERVFAIL reply from 10.11.1.20, trying next server
Server:		10.11.1.19
Address:	10.11.1.19#53

** server can't find en: SERVFAIL

```

**After adding `search wikipedia.org` to `/etc/resolv.conf`:**
```
> dns.resolve4('en', {ttl: true, search: true}, (err, result) => { if (err) console.error(err); console.log(result); })
QueryReqWrap {
  bindingName: 'queryA',
  callback: [Function],
  hostname: 'en',
  oncomplete: [Function: onresolve],
  ttl: true,
  search: true,
  domain:
   Domain {
     domain: null,
     _events:
      { removeListener: [Function: updateExceptionCapture],
        newListener: [Function: updateExceptionCapture],
        error: [Function: debugDomainError] },
     _eventsCount: 3,
     _maxListeners: undefined,
     members: [] },
  channel:
   ChannelWrap {
     domain:
      Domain {
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        members: [] } } }
> [ { address: '103.102.166.224', ttl: 228 } ]
```

I can verify this against `nslookup`:
```
$ nslookup "en"
Server:		10.11.1.20
Address:	10.11.1.20#53

Non-authoritative answer:
Name:	en.wikipedia.org
Address: 103.102.166.224
```

Reply:
I think I'm coming across a weird behaviour too, not sure if this is a bug though.

Steps to reproduce:
1. Start `./out/Release/node`, the build result of [this PR's branch](https://github.com/SirR4T/node/tree/fix19548-support-cares-search)
2. Require dns module.
3. launch this query for `"en"`: `dns.resolve4('en', {ttl: true, search: true}, (err, result) => { if (err) console.error(err); console.log(result); })`
4. Observe that `ESERVFAIL` error is thrown (this is expected).
5. In a different shell, update `/etc/resolv.conf` to have this line at the end: `search wikipedia.org`
6. Switch back to original shell, relaunch the same query
7. Observe that `ESERVFAIL` is still thrown.
8. Shutdown the repl and start again
9. Relaunch the query, and observe that it now works!

I expect that the `dns.resolve4()` query should work with the latest changes in the `/etc/resolv.conf` file, but I suspect it is getting cached somewhere, during startup / require.